---
title: "Pooling RNA-seq and Assembling Models"
author: "Peng Liu, Colin N. Dewey, and Sündüz Keleş"
date: "`r Sys.Date()`"
output: 
    rmarkdown::html_vignette:
        toc: true
        fig_caption: yes
vignette: >
  %\VignetteIndexEntry{Pooling RNA-seq and Assembling Models}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


## Introduction

Pooling RNA-seq and Assembling Models (__PRAM__) utilizes multiple RNA-seq 
datasets to predict transcript models.  The workflow of PRAM contains five 
steps, which is shown in 
the figure below with function names and associated key parameters.  In later
sections of this vignette, we will describe each function in details.

```{r, fig.cap='PRAM workflow', out.width='579px', out.height='324px', echo=F}
knitr::include_graphics("mainPRAMWorkflow.jpg")
```

## Installation

```
devtools::install_github('https://github.com/pliu55/pram')
```

## Quick start

### Predict transcript models only
```
runPRAM(in_gtf, in_bamv, out_gtf)
```

- `in_gtf`:  An input GTF file defining genomic coordinates of existing genes. 
             Required to have `gene_id` in the attribute column (column 9).
- `in_bamv`:  A vector of input BAM file(s) of RNA-seq alignments
- `out_gtf`:  An output GTF file of predicted transcript models


### Predict transcript models and screen them by ChIP-seq data
```
runPRAM(in_gtf, in_bamv, out_gtf, in_bedv, training_gtf, training_tpms, cufflinks)
```

- `in_gtf`, `in_bamv`, `out_gtf`, and `cufflinks` are the same as described
   [above](#predict-transcript-models-only)
- `in_bedv`:  A vector of BED file(s) of ChIP-seq alignments.
- `training_gtf`:  A GTF file defining genomic coordinates of training 
                   transcripts 
- `training_tpms`:  A vector of RSEM quantification results for training
                    transcripts



## Define intergenic genomic ranges: `defIgRanges()`

### Example

## Prepare input RNA-seq alignments: `prepIgBam()`

## Build transcript models: `buildModel()`

### Example

### Prediction methods
PRAM has implemented seven transcript reconstruction methods.  Five methods, 
__plcf__, __plst__, __cfmg__, __cftc__, and __stmg__, are meta-assembly methods 
that utilize multiple RNA-seq datasets to predict a single set of 
transcript models.  The other two methods, __cf__ and __st__, predicts
transcript models from a single RNA-seq dataset.  The table below compares
the prediction steps for these seven methods.  By default, `buildModel()` uses
__plcf__ to predict transcript models.  Users can use the `method` argument in
`buildModel()` to specifiy their preferred prediction method.

Table: Prediction steps of PRAM's seven methods

| method | preparing RNA-seq input  | building transcripts | assembling transcripts |
|:------:|:------------------------:|:---------------:|:-----------------:|
| plcf   | pooling alignments       | Cufflinks       | no                |
| plst   | pooling alignments       | StringTie       | no                |
| cfmg   | no                       | Cufflinks       | Cuffmerge         |
| cftc   | no                       | Cufflinks       | TACO              |
| stmg   | no                       | StringTie       | StringTie-merge   |
| cf     | no                       | Cufflinks       | no                |
| st     | no                       | StringTie       | no                |


### Required external software
Depending on the ternal software to reconstruct transcript models. 

- `cufflinks`: Cufflinks executable, which can be downloaded for 
  [Linux](http://cole-trapnell-lab.github.io/cufflinks/assets/downloads/cufflinks-2.2.1.Linux_x86_64.tar.gz) 
  or [MacOS](http://cole-trapnell-lab.github.io/cufflinks/assets/downloads/cufflinks-2.2.1.OSX_x86_64.tar.gz)


## Select transcript models: `selModel()`

## Screen transcript models: `screenModel()`

Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style

## Vignette Info

Note the various macros within the `vignette` section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
